File: manufacturing_customizations/customizations/material_request.py
import frappe
from frappe.model.document import Document

@frappe.whitelist()
def create_purchase_orders(material_request):
    material_request_doc = frappe.get_doc('Material Request', material_request)
    items_by_supplier = {}

    # Group items by supplier
    for item in material_request_doc.items:
        default_suppliers = frappe.get_all('Default Supplier', filters={'parent': item.item_code}, fields=['default_suppliers'])
        for entry in default_suppliers:
            supplier = entry.default_suppliers
            if supplier:
                if supplier not in items_by_supplier:
                    items_by_supplier[supplier] = []
                items_by_supplier[supplier].append(item)

    # Create Purchase Orders for each supplier
    purchase_orders = []
    existing_orders = []
    for supplier, items in items_by_supplier.items():
        existing_po = frappe.get_all('Purchase Order', filters={
            'material_request': material_request,
            'supplier': supplier,
            'docstatus': ['<', 2]  # Check for draft and submitted POs
        }, fields=['name'])

        if existing_po:
            existing_orders.append({'supplier': supplier, 'po_name': existing_po[0].name})
            continue 

        po = frappe.new_doc('Purchase Order')
        po.supplier = supplier
        for item in items:
            po.append('items', {
                'item_code': item.item_code,
                'qty': item.qty,
                'schedule_date': item.schedule_date,
                'rate': item.rate,
                'warehouse': item.warehouse,
                'material_request': material_request,
                'material_request_item': item.name
            })
        po.insert()
        purchase_orders.append({'name': po.name, 'supplier': po.supplier})

    return {
        'created': purchase_orders,
        'existing': existing_orders
    }

@frappe.whitelist()
def get_suppliers_for_item(item_code):
    default_suppliers = frappe.get_all('Default Supplier', filters={'parent': item_code}, fields=['default_suppliers'])
    supplier_list = [entry.default_suppliers for entry in default_suppliers]
    return supplier_list

File: manufacturing_customizations/public/js/material_request.js
frappe.ui.form.on('Material Request', {
    refresh: function(frm) {
        if (!frm.doc.__islocal && frm.doc.docstatus === 1) {
            frm.add_custom_button(__('Auto Create PO'), function() {
                frappe.call({
                    method: 'manufacturing_customizations.customizations.material_request.create_purchase_orders',
                    args: {
                        material_request: frm.doc.name
                    },
                    callback: function(response) {
                        let message = '';
                        if (response.message.created.length > 0) {
                            message += '<b>' + __('Purchase Orders Created Successfully:') + '</b><br>';
                            response.message.created.forEach(po => {
                                message += `<a href='/app/purchase-order/${po.name}'>${po.name}</a> (Supplier: ${po.supplier})<br>`;
                            });
                        }
                        if (response.message.existing.length > 0) {
                            message += '<br><b>' + __('Existing Purchase Orders:') + '</b><br>';
                            response.message.existing.forEach(po => {
                                message += `A purchase order for supplier ${po.supplier} has already been created: <a href='/app/purchase-order/${po.po_name}'>${po.po_name}</a><br>`;
                            });
                        }
                        if (response.message.created.length === 0 && response.message.existing.length > 0) {
                            message = __('All purchase orders for the suppliers have already been created.');
                        }
                        frappe.msgprint({
                            title: __('Purchase Order Creation'),
                            indicator: 'green',
                            message: message
                        });
                    }
                });
            });
        }
    }
});
frappe.ui.form.on('Material Request Item', {
    item_code: function(frm, cdt, cdn) {
        const row = locals[cdt][cdn];
        if (row.item_code) {
            frappe.call({
                method: 'manufacturing_customizations.customizations.material_request.get_suppliers_for_item',
                args: {
                    item_code: row.item_code
                },
                callback: function(response) {
                    if (response.message) {
                        const suppliers = response.message.join(', ');
                        frappe.model.set_value(cdt, cdn, 'custom_default_suppliers', suppliers);
                    }
                }
            });
        }
    }
});

frappe.ui.form.on('Material Request', {
    refresh: function(frm) {
        if (!frm.fields_dict.items) return;

        frm.fields_dict.items.grid.add_custom_button(__('Check Item Info'), function() {
            check_item_info(frm);
        });

        refresh_items_info(frm);
    },
});

function refresh_items_info(frm) {
    let promises = frm.doc.items.map(function(item) {
        return new Promise((resolve) => {
            frappe.call({
                method: 'manufacturing_customizations.customizations.material_request.get_suppliers_for_item',
                args: {
                    item_code: item.item_code
                },
                callback: function(response) {
                    if (response.message) {
                        const suppliers = response.message.join(', ');
                        frappe.model.set_value(item.doctype, item.name, 'custom_default_suppliers', suppliers);
                    }
                    resolve();
                }
            });
        });
    });

    return Promise.all(promises);
}

function check_item_info(frm) {
    refresh_items_info(frm).then(() => {
        let items_info = [];

        frm.doc.items.forEach(function(item, index) {
            items_info.push({
                index: index + 1,
                item_code: item.item_code,
                item_name: item.item_name,
                custom_default_suppliers: item.custom_default_suppliers || ''
            });
        });

        // Sort items by their original order (index)
        items_info.sort((a, b) => a.index - b.index);

        let message = `
            <table class="table table-bordered">
                <thead>
                    <tr>
                        <th>${__('Sr')}</th>
                        <th>${__('Item Code')}</th>
                        <th>${__('Item Name')}</th>
                        <th>${__('Custom Default Suppliers')}</th>
                    </tr>
                </thead>
                <tbody>`;

        items_info.forEach(function(item) {
            let row_style = item.custom_default_suppliers ? '' : ' style="color: red;"';
            message += `
                <tr${row_style}>
                    <td>${item.index}</td>
                    <td>${item.item_code}</td>
                    <td>${item.item_name}</td>
                    <td>${item.custom_default_suppliers}</td>
                </tr>`;
        });

        message += `
                </tbody>
            </table>`;

        frappe.msgprint({
            title: __('Item Information'),
            indicator: 'blue',
            message: message
        });
    });
}




